src/NNLSSolverInterface.jl

```diff
@@ -0,0 +1,102 @@
+module NNLSSolverInterface
+
+export QPSolver
+
+using NNLS: QPWorkspace, load!, solve!
+importall MathProgBase.SolverInterface
+
+immutable QPSolver <: AbstractMathProgSolver
+    workspace::QPWorkspace{Float64, Int}
+
+    QPSolver() = new(QPWorkspace{Float64, Int}(0, 0))
+end
+
+type QPModel <: AbstractLinearQuadraticModel
+    workspace::QPWorkspace{Float64, Int}
+    lb::Vector{Float64}
+    ub::Vector{Float64}
+    A::Matrix{Float64}
+    constr_lb::Vector{Float64}
+    constr_ub::Vector{Float64}
+    Q::Matrix{Float64}
+    q::Vector{Float64}
+    sense::Symbol
+    solution::Vector{Float64}
+    duals::Vector{Float64}
+
+    QPModel(work::QPWorkspace) = new(work)
+end
+
+LinearQuadraticModel(s::QPSolver) = QPModel(s.workspace)
+
+numvar(work::QPWorkspace) = size(work.G, 2)
+numconstr(work::QPWorkspace) = size(work.G, 1)
+
+numvar(m::QPModel) = numvar(m.workspace)
+numconstr(m::QPModel) = numconstr(m.workspace)
+
+function loadproblem!(m::QPModel, A, lb, ub, q, constr_lb, constr_ub, sense)
```

Comment:
I think it's fine the way it is for now, but I'm not sure if I agree with that last statement. Isn't `loadproblem!` going to be called in a loop?

So currently [`MathProgBase.quadprog`](https://github.com/JuliaOpt/MathProgBase.jl/blob/bdaaa744d0ef43e95dfd6b2f3eea9c8fbd6e2a9d/src/HighLevelInterface/quadprog.jl#L20) just creates a new `LinearQuadraticModel`, so the point is moot if you're using `quadprog`. But I think that if all of the fields of `NNLSModel` moved to `NNLSSolver`, `quadprog` would actually not allocate apart from the `QuadprogSolution` stuff at the end of the function. However, I see your point with the two different models.

If you're using JuMP, `loadproblem!` is going to be called on the `Model`'s internal model in [`build`](https://github.com/JuliaOpt/JuMP.jl/blob/8ea959a5cb9bb95b554b5636b1d2e21ad25c5c2b/src/solvers.jl#L338), which is called from `solve`. So I don't think it's possible to solve a problem without calling `loadproblem!`, and so the allocation does matter. It's probably a tiny amount compared to other allocations happening in JuMP, but if we eventually want a true zero-allocation solver, I think `loadproblem!` needs to not allocate either.

Can't `loadproblem!` just `resize!` everything *if necessary* based on the method arguments? It's OK if `loadproblem!` allocates the first time for a given number of variables and constraints of course.

BTW, the LAPACK calls in the QP --> NNLS conversion allocate a tiny amount (some stupid info `Ref`s passed to LAPACK), so we're not going to be completely allocation-free right now anyway.